src/codegen/generators/expression_generator.rs

```diff
@@ -319,6 +273,47 @@ impl<'a, 'b> ExpressionCodeGenerator<'a, 'b> {
         }
     }
 
+    pub fn generate_direct_access_index(
+        &self,
+        access: &DirectAccessType,
+        index: &AstStatement,
+        access_type: &DataTypeInformation,
+        target_type: &DataType,
+    ) -> Result<IntValue<'a>, CompileError> {
+        let reference = self.generate_expression(index)?;
+        //Load the reference
+        if reference.is_int_value() {
+            //TODO why is this cast necessary???
```

Comment:
I think we can remove this cast if we set the type hint on direct access to the type of the previous segment in the qualified expression chain.
I can do that in a separate commit.